[Issue 1907] TO API for backup edge cachegroup

Vijay-1 opened the pull request:
[Issue 1907] Backup edge cache group TO API / TR cr-config changes

https://github.com/apache/incubator-trafficcontrol/issues/1907

Contains the following changes:
Traffic Router - Backup edge configuration parsing and decision making code changes
Traffic Ops - Mojolicious Perl CRUD APIs

This is the initial version of code. Only minimal testing has been done. Will run more cases and update. Hosting it in advance to get review comments :-)

asfgit commented:
Can one of the admins verify this patch?

asfgit commented:
Can one of the admins verify this patch?

dangogh commented:
@Vijay-1 can you change the title of this PR so it's more descriptive of the actual change?  A link in the title doesn't work well..   Something like this is good:
`[Issue 1907] TO API for backup edge cache group`

limited commented on CHANGELOG.md:
```diff
@@ -34,6 +34,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
   - /api/1.3/system/info
   - /api/1.3/types (R)
 - Fair Queuing Pacing: Using the FQ Pacing Rate parameter in Delivery Services allows operators to limit the rate of individual sessions to the edge cache. This feature requires a Trafficserver RPM containing the fq_pacing experimental plugin AND setting 'fq' as the default Linux qdisc in sysctl. 
+- Backup Edge Cache group: Backup Edge group for a particular cache group can be configured using Traffic Router's cr-config.json configuration file. With this, users will be able control traffic within portions of their network, thereby avoiding choosing fallback cache groups using geo coordinates(getClosestAvailableCachegroup). This can be controlled using "backupLocations" which contains confi ...
```
Since the changelog is user visible, its better to describe the changes as configured using Traffic Ops API. You could also make the description more concise as you have more detailed documentation elsewhere. 

Consider something along the lines of:
```
  Configure a backup list of edge cache groups through the Traffic Ops API. A Backup edge cache group is used if the matched group in the CZF is not available. If the backup edge cache groups are not available, geo location can optionally be used as a further backup. See APIs here...
```

Vijay-1 replied:
Will do.

mitchell852 commented on docs/source/development/traffic_ops_api/v12/cachegroup_fallbacks.rst:
```diff
@@ -0,0 +1,266 @@
+
+**PUT /api/1.2/cachegroups/:id/fallbacks**
+
+  Create cache group.
```
can you add a description of what this route is used for. It is not clear to me what a PUT does. Does it replace all existing fallbacks?

mitchell852 commented on docs/source/development/traffic_ops_api/v12/cachegroup_fallbacks.rst:
```diff
@@ -0,0 +1,266 @@
+
+**DELETE /api/1.2/cachegroups/{:id}/fallbacks**
+
+  Delete cache group. The request to delete a cache group, which has servers or child cache group, will be rejected.
```
Need a real description here.

mitchell852 commented on traffic_ops/app/lib/API/Cachegroup.pm:
```diff
@@ -57,6 +57,7 @@ sub index {
 				"lastUpdated"                   => $row->last_updated,
 				"parentCachegroupId"            => $row->parent_cachegroup_id,
 				"parentCachegroupName"          => ( defined $row->parent_cachegroup_id ) ? $idnames{ $row->parent_cachegroup_id } : undef,
+				"fallback_to_closest"           => \$row->fallback_to_closest,
```
API responses need to be camelcase

DylanVolz commented on docs/source/development/traffic_ops_api/v12/cachegroup_fallbacks.rst:
```diff
@@ -0,0 +1,266 @@
+              "text": "Backup configuration for 51 successful."
+            }
+         ],
+       "response": {
```
+1, the extra keys will make client code unnecessarily complex.

Vijay-1 commented:
@mitchell852 
I have addressed your comments on the API Response structure.

I have a comment on one of your comments:
>>PUT should be for updating a specific fallback so the route should be PUT >>/api/1.2/cachegroups/:id/fallbacks/:fb_id

In this we dont have a specific fallback (i.e) no unique id associated with fallback. This configuration is always tied with a cache group id and hence cache group id alone is enough.

mitchell852 commented:
@Vijay-1 - can you just do 3 endpoints? I still dont' understand the need for the PUT.

GET /api/cachegroups/:id/fallbacks <-- return an array of fallbacks for the cachegroup
POST /api/cachegroups/:id/fallbacks <-- create fallbacks for a cachegroup. so the json request is an  array of fallbacks
DELETE /api/cg/:id/fallbacks <-- delete all fallbacks for the cachegroup

mitchell852 commented on docs/source/development/traffic_ops_api/v12/cachegroup_fallbacks.rst:
```diff
@@ -0,0 +1,266 @@
+
+**PUT /api/1.2/cachegroups/:id/fallbacks**
+
+  Updates an already existing fallback list for the cache group (Existing fallback list will be deleted). 
```
why don't you delete this route and just use POST instead where POST always creates fallbacks and removes the current ones if there are any.

Vijay-1 replied:
May be i should reword that sentence like the following:

Updates / replaces an already existing fallback list for the cache group.

Vijay-1 replied:
Ok. Makes sense. I agree that we can do away with PUT. Hope every one is fine with this.

DylanVolz commented:
After looking over how this endpoint can be written and go, and our plan for api design moving forward (https://cwiki.apache.org/confluence/display/TC/API+Guidelines) I would suggest a few changes:

Change the route to not be nested. We are moving away from nested routes to remove ambiguity related to keys and simplify routing concerns.
A good example is @mitchell852 's recent PR for comments on Delivery Service Requests. https://github.com/apache/incubator-trafficcontrol/pull/2027/files#diff-e1d71774f8ad89e82bd76378c68555e3R117
So instead of "/api/$version/cachegroups/:id/fallbacks" the route could be formed: "/api/$version/cachegroup_fallbacks" and the cachegroup id would be provided as a query parameter.

Change the api response to include the cachegroup id in each fallback object in the array. This allows for more options for filtering and more accurately represents the data in the database.

Change the name key to something more descriptive / tied to the fallback cachegroup to make it clear.

With these changes the GET request would change from:

```
GET /api/1.3/cachegroups/1/fallbacks

	{
       "response": [
          {
                "order":10,
                "name":"GROUP2"
          },
          {
          	"order":11,
          	"name":"GROUP3"
          }
       ]
    }
```

  to:

```
  GET /api/1.3/cachegroup_fallbacks?cachegroupID=1

  	{
       "response": [
          {
               "cachegroupID":1
               "order":10,
               "fallbackName":"GROUP2"
          },
          {
                "cachegroupID":1
          	"order":11,
          	"fallbackName":"GROUP3"
          }
       ]
    }

```
   

These changes will allow the api to easily use some prebuilt components and will give us some additional functionality for free in go, such as being able to filter by the fallback group's name and get all the cachegroups that fall back to it for example.

Vijay-1 commented:
In perl/ Mojo, i can see this query params only for indexes, where we are fetching an entire set and ordering , say for example all available cache groups/ delivery services etc and ordering them by id / name etc.
But in this case we always go by a specific cache id. When i fetch data for /api/cachegroup_fallbacks?cgid=1, i have to return results in a particular format and for ?fbid, i have to pack results in particular format.

Say for example:
for /cachegroup_fallbacks?fallbackId=5

GET response may be a list of cache groups which has server 5 as backup
{"response":{"fallbackID":"5","list":["GROUP1","GROUP2"]}}


whereas /cachegroup_fallbacks?cacheGroupId=5
Backup configuration for server 5 along with its order
{"response":[{"cacheGroupId":"1","order":9,"name":"GROUP6"},{"cacheGroupId":"1","order":10,"name":"GROUP2"}]}

is that what you meant?

I dont think current Perl/ Mojo implementation has any thing of this sort. 

AFIK, the only point i see in supporting these APIs is the similarity it may share with GO APIs. Is this alone is enough to change nested routes to the ones which supports query params? Why cant we leave this as such for Perl / Mojo?

hongfeiweb commented:
@DylanVolz - i am not convinced that moving the URL to top level (as cachegroup_fallbacks) for this functionality is the right direction. I think the guideline wiki cited a good use of toplevel for profile & parameters. In such case, the parameters and profiles are loosely associated where params don't necessarily get associated with any profile. The fallbacks however are different in that it's just some additional, optional attributes to an otherwise regular cachgroup therefore should be accessible only through ../cachgroup/... hierarchy  ihmo. I don't think the database storage normalization has to necessarily influence the API representation.

mitchell852 commented:
@hongfeiweb - we're trying to move away from "route" or "path" parameters to remove some of the ambiguity that they have caused. Also, to condense our number of routes.

For example, we have routes like this: 

GET /api/servers/{id}
GET /api/servers/{hostname}

We'd rather just have one route that supports multiple query params.

GET /api/servers?id=4 or GET /api/servers?hostname=foo-bar

It's not that I'm opposed to nested routes per se, but it's hard to have nested routes if you eliminate route/path parameters.

I've updated the TO API guidelines to reflect our new approach to APIs as we move forward with Golang:

https://cwiki.apache.org/confluence/display/TC/API+Guidelines

Vijay-1 commented:
@DylanVolz 
I have tried to address your comments. Request your review on the same.

rawlinp commented:
@Vijay-1 - I just wanted to make sure you were aware that the cachegroup API was recently rewritten in Go if you wanted to make your changes there rather than in Perl. As of #2080 merging, the Perl cachegroup routes are no longer used.

DylanVolz commented:
@mitchell852 I think this looks good now, I verified it works as expected. The only outstanding issue is the merging of #2080 that @rawlinp pointed out. Until that is updated this feature won't be able to be used, but I am unsure that blocks this PR necessarily.

mitchell852 commented:
@rawlinp @rivasj - can you guys look at the changes made to TR?

mitchell852 commented on traffic_ops/app/lib/API/Cachegroup.pm:
```diff
@@ -164,6 +166,7 @@ sub update {
 		latitude                       => $params->{latitude},
 		longitude                      => $params->{longitude},
 		parent_cachegroup_id           => $params->{parentCachegroupId},
+		fallback_to_closest            => $params->{fallbackToClosest},
```
Because this field was not added to Traffic Portal, users of the TP will always be setting this to null when they update a cache group. Is that OK? 

I think you need to add this field to the Traffic Portal - form.cacheGroup.tpl.html

mitchell852 replied:
fallbackToClosest needs to be added to the Traffic Portal

form.cacheGroup.tpl.html

mitchell852 commented:
@Vijay-1 - I think you need to rethink your design of this feature. The way you have it built will be very difficult to incorporate into the Traffic Portal which uses POSTs for create only and PUTs for update. Here are my suggestions:

GET /cachegroup_fallbacks[?id=|?cachegroupId=|?fallbackId=]

- Get all cachegroup fallbacks
- Get a specific cachegroup fallback
- Get all cachegroup fallbacks for a cg
- Get all cgs for a fallback

POST /cachegroup_fallbacks [ { "cacheGroupId": 1, "fallbackId": 27, "fallbackOrder": 10 }, {...} ]

- create one or more fallbacks (no deleting here)
- no query param on this route. use the cacheGroupId param to determine which cachegroup to create fallback for.

PUT /cachegroup_fallbacks[?id=]

- update a cachegroup fallback by id

DELETE /cachegroup_fallbacks[?id=|cachegroupId=]

- delete a specific fallback
- delete all fallbacks for a cachegroup

Vijay-1 commented:
@mitchell852 
Also i think the way cachegroup_fallbacks is designed , this table has no id.  So I am not sure on the query parameter id on GET.

It should be either cacheGroupId or fallbackId. Both Ids are actually cache group Ids(cachegroup Table). Since cachegroup_fallbacks is always dependent on cachegroup only we can have the following GET API endpoints.
Get a specific cachegroup's fallback (?cacheGroupId)
Get all cachegroups which has a specific fallback (?fallbackId)

Vijay-1 commented:
@mitchell852 

Some minot modification from your suggestions. Request your views before actually modifying those.
fetch list 
    Cache group for a fallbackId
        Example: GET api/{version}/cachegroup_fallbacks?fallbackId=6
    Fallbacks for a cacheGroupId
         Example: GET /api/{version}/cachegroup_fallbacks?cacheGroupId=1

delete one
    DELETE all the fallbacks for a cache
         Example: DELETE api/{Version}/cachegroup_fallbacks?cacheGroupId=1
    DELETE a specific fallback for a cache
         Example: DELETE api/{Version}/cachegroup_fallbacks?cacheGroupId=1&fallbackId=5"

create one
       POST /api/{version}/cachegroup_fallbacks?cacheGroupId=1";
       POST /api/{version}/cachegroup_fallbacks";
      
       POST will update if an already existing cache:fallback combination is sent in . if no such combination exists it will create one   

Most of this functionalities is already there except for POST's deletion and Delete's specific row requirement.

mitchell852 commented:
@Vijay-1 - i think that makes sense

fetch list
Cache group for a fallbackId
Example: GET api/{version}/cachegroup_fallbacks?fallbackId=6 returns
[ { cachegroupId: 3, cachegroupName: 'cg-3', fallbackId: 6, fallbackName: 'fb-6' }, { cachegroupId: 4, cachegroupName: 'cg-4', fallbackId: 6, fallbackName: 'fb-6' }  ]

Fallbacks for a cacheGroupId
Example: GET /api/{version}/cachegroup_fallbacks?cacheGroupId=3 returns
[ { cachegroupId: 3, cachegroupName: 'cg-3', fallbackId: 6, fallbackName: 'fb-6' }, { cachegroupId: 3, cachegroupName: 'cg-3', fallbackId: 7, fallbackName: 'fb-7' }  ]

^^ although the cg is not needed in the response, it keeps the structure the same regardless of which query param you send in.

delete one
DELETE all the fallbacks for a cache
Example: DELETE api/{Version}/cachegroup_fallbacks?cacheGroupId=1
DELETE a specific fallback for a cache
Example: DELETE api/{Version}/cachegroup_fallbacks?cacheGroupId=1&fallbackId=5"

but for POST (create), why not just one route like this:

POST /api/{version}/cachegroup_fallbacks where the request is an array of one or more that looks like this:

[ { cgId: 3, fbId: 6 }, { cgId: 3, fbId: 7 }, { cgId: 3, fbId: 8 }  ]

^^ insert a row for each item in the array. if it is already in the database, skip it.

Vijay-1 commented:
@mitchell852 
Looks like you missed the Order parameter:
PUT Request 
    [ { "cacheGroupId": 1, "fallbackName": "GROUP2", "fallbackOrder": 10 } ]

>> fallbackName instead of fallback Id is just a convenience, so that it will be use full while debugging :grinning: 

GET Response
    { "response": [ { "cacheGroupId":1, "cacheGroupName":"GROUP1", "fallbackId":2, "fallbackOrder":10, "fallbackName":"GROUP2" } ] }

Also POST , updates rather than skipping , because there may be a change in Order

Vijay-1 commented:
@mitchell852 

I spent some time in the GoLang version implementation today and i have some doubts on GoLang version's flexibility over Perl.

Request you to point me to existing GoLang implementation for the cases described below which will help me in implementing GoLang version of Endpoints

Given below are my observations:

DELETE:

With GoLang, looks like we can have either one of the APIs and not both. Some of the modules i explored in GoLang version, I see only one format of delete query. If yes, should we stick with #2?

1. DELETE all the fallbacks for a cache
    Example: DELETE api/{Version}/cachegroup_fallbacks?cacheGroupId=1

2. DELETE a specific fallback for a cache
    Example: DELETE api/{Version}/cachegroup_fallbacks?cacheGroupId=1&fallbackId=5"
===================================================================
GET:
I am not sure whether we can return the following for GET

 { "cacheGroupId":1, "cacheGroupName":"GROUP1", "fallbackId":2, "fallbackOrder":10, "fallbackName":"GROUP2" }

 Since CacheGroupName and fallbackName is not cachegroup_fallback's column (cachegroup's column), i am not sure we can return them in GoLang version.
===============================================================

POST:

Can you point me to a GoLang module which handles json array.

This works fine for me:
 my $json = '{"primaryId":3, "backupId": 1, "setOrder": 142}';

But when i use arrays like the following, i am seeing the following error:
>>>>cannot unmarshal array into Go value of type cachegroupfallback.TOCACHEGROUPFALLBACK]

 my $json = '[{"primaryId":3, "backupId": 1, "setOrder": 142}]';

Post only for creates
===================================================================
 PUT:
 I belive we may have to support PUT /cachegroup_fallbacks[?cachegroupId=4&fallbackId=6]

PUT for updates
===================================================================

rawlinp commented:
I reviewed the TR-side of the PR which looks good to me. Once the API-side is solid I'll pull this down and try out the whole thing.

mitchell852 commented on docs/source/development/traffic_ops_api/v12/cachegroup_fallbacks.rst:
```diff
@@ -0,0 +1,209 @@
+
+  **Request Query Parameters**
+
+  Query parameter is mandatory. Either one of the parameters must be used. Both cant be used simultaneously.
```
Can i send through both query params to fetch ONE record? the ability to fetch one record is something all our other endpoints support. I.e.

GET /api/1.2/cachegroup_fallbacks?cacheGroupId=&fallbackId=

mitchell852 commented on docs/source/development/traffic_ops_api/v12/cachegroup_fallbacks.rst:
```diff
@@ -0,0 +1,209 @@
+
+**POST /api/1.2/cachegroup_fallbacks**
+
+  Creates/ Updates fallback list for the cache group.
```
Can you add a little more detail here. How does this work exactly? Send in an array to the POST. for each item in the array, if doesn't exist, create it. if exists, update it?

mitchell852 commented on docs/source/development/traffic_ops_api/v12/cachegroup_fallbacks.rst:
```diff
@@ -0,0 +1,209 @@
+
+**DELETE /api/1.2/cachegroup_fallbacks**
+
+  Delete fallback list assigned to the cache group.
```
You can also delete one, right? Can you update the docs to reflect that? For example:

DELETE /api/1.2/cachegroup_fallbacks?cacheGroupId=3&fallbackId=23

mitchell852 commented on traffic_ops/app/lib/API/CachegroupFallback.pm:
```diff
@@ -0,0 +1,208 @@
+use MojoPlugins::Response;
+use Validate::Tiny ':all';
+
+sub delete {
```
you should be able to delete only one fallback, right? i don't see that ability to do that.

mitchell852 commented on traffic_ops/app/lib/API/CachegroupFallback.pm:
```diff
@@ -0,0 +1,208 @@
+	}
+}
+
+sub show {
```
how can i fetch just one row? like this

GET /api/1.2/cachegroup_fallbacks?cacheGroupId=&fallbackId=

mitchell852 commented on traffic_ops/app/lib/API/CachegroupFallback.pm:
```diff
@@ -0,0 +1,208 @@
+	my $rs_backups = $self->db->resultset('CachegroupFallback')->search( { primary_cg => $cache_id } );
+	if ( defined ($rs_backups->next) ) {
+		&log( $self, "Deleting all existing backup configuration for cache group $cache_id");
+		$rs_backups->delete();
```
this is deleting the entire list and replacing it with the contents of the POST

i thought we discussed simply adding the item if it doesn't exist in the db and updating it if it does.

mitchell852 commented:
@Vijay-1 and I discussed these routes but I don't think the changes have been made yet to support them:

GET /api/1.2/cachegroup_fallbacks?cacheGroupId= <-- get all fallbacks for a cg
GET /api/1.2/cachegroup_fallbacks?fallbackId= <-- get all cg's for a fb
GET /api/1.2/cachegroup_fallbacks?cacheGroupId=&fallbackId= <-- get one cg fb

POST /api/1.2/cachegroup_fallbacks <-- create one or more fallbacks. if a fallback already exists, report that it already exists and bail (403)

PUT /api/1.2/cachegroup_fallbacks <-- update one or more fallbacks, skip if fallback doesn't exist or fallback is invalid

DELETE /api/1.2/cachegroup_fallbacks?cacheGroupId=3&fallbackId=23 <-- delete one
DELETE /api/1.2/cachegroup_fallbacks?cacheGroupId=3 <-- delete all fallbacks for cg3
DELETE /api/1.2/cachegroup_fallbacks?fallbackId=23 <-- delete all cachegroups for fb23

Vijay-1 commented:
@mitchell852 @elsloo 
Will upload the changes in couple of days. Caught up in some thing else.

mitchell852 commented on traffic_ops/app/lib/API/CachegroupFallback.pm:
```diff
@@ -0,0 +1,283 @@
+
+	my $fallback_config = $self->db->resultset('CachegroupFallback')->find({ primary_cg => $cache_id});
+
+	if ( defined ($fallback_config) ) {
```
I don't  understand why this is here? how can i add more fallbacks for a cachegroup? if there is 1 fallback in place, this will always error out.

Vijay-1 replied:
@mitchell852 

Yes. I belive, that is what we have discussed. If one has to update , he has to PUT instead of POST.
POST more than once result in 400 Bad request.

PUT a not POST-ed resource will result in 404 Not found

mitchell852 replied:
ok

mitchell852 commented on traffic_ops/app/lib/API/CachegroupFallback.pm:
```diff
@@ -0,0 +1,283 @@
+		$rs_backups = $self->db->resultset('CachegroupFallback')->search({ backup_cg => $fallback_id}, {order_by => 'set_order'});
+	}
+
+	if ( defined ($rs_backups) && ($rs_backups->count > 0) ) {
```
just return an empty array if count=0. don't return a 404 ($self->not_found())

Vijay-1 replied:
I am not sure on why we shouldn't return 404. To me it looks like a valid behavior since we cant find records for such a query.

mitchell852 replied:
to ensure we have a consistent API, you need to follow the examples of the other apis. If you ask for a list of resources and there are none, an empty array is returned instead of a 404

for example, 

GET /api/foos?active=true - if there are no active foos, then [] is returned and not 404.

404 implies an invalid resource path. /api/foos?active=true is a valid resource path, there just happens to be no active foos.

this on the other  hand would result in a 404:

GET /api/foos/777/bars <-- in this case there is no foo with id=777, so 404 would be valid.

Vijay-1 replied:
Ok. Will address this.

mitchell852 commented:
I still don't agree with how the POST and PUT works. In my  opinion, a POST should add fallbacks and a PUT should update existing fallbacks. This is how the rest of the API works.

For example:

POST /api/1.2/cachegroup_fallbacks [ fallback11, fallback12 ] <-- if there are already 10 fallbacks for cg1, add 2 more to the list making 12 fallbacks for cg1

PUT /api/1.2/cachegroup_fallbacks [ fallback11, fallback12 ] <-- go find fallback11 and fallback12 and update them

Vijay-1 commented:
what happens for the following cases:
POST /api/1.2/cachegroup_fallbacks [ fallback11, fallback12 ] 

followed by 
POST /api/1.2/cachegroup_fallbacks [ fallback11, fallback12 ]  ==> do nothing?
and
POST /api/1.2/cachegroup_fallbacks [ fallback11, fallback12,fallback13 ] ==> create only fallback13?

====================

PUT /api/1.2/cachegroup_fallbacks [ fallback11, fallback12 ] 

Followed by 
PUT /api/1.2/cachegroup_fallbacks [ fallback11, fallback12,fallback14 ] <== update only fallback11 and fallback12 and skip fallback14?

mitchell852 commented:
this seems ok from an api perspective. need someone to review/test the TR part.

@rawlinp @rivasj @elsloo @ajschmidt any takers?

mitchell852 commented on traffic_ops/app/lib/UI/Topology.pm:
```diff
@@ -250,6 +250,15 @@ sub gen_crconfig_json {
                 $data_obj->{'edgeLocations'}->{ $row->cachegroup->name }->{'longitude'} = $row->cachegroup->longitude + 0;
+                $data_obj->{'edgeLocations'}->{ $row->cachegroup->name }->{'backupLocations'}->{'fallbackToClosest'} = $row->cachegroup->fallback_to_closest ? "true" : "false";
+
+                my $rs_backups = $self->db->resultset('CachegroupFallback')->search({ primary_cg => $row->cachegroup->id}, {order_by => 'set_order'});
```
gen_crconfig_json was rewritten in Golang so this logic will need to be added there as well otherwise it will be lost. You might want to reach out to @rob05c

mitchell852 replied:
@Vijay-1 - this can't be merged until this is addressed.

mitchell852 commented on traffic_ops/app/db/migrations/20180321000000_cache_group_fallback.sql:
```diff
@@ -0,0 +1,38 @@
+/*
+
```
this migration won't work. the date is too  old.

Vijay-1 replied:
Renamed.

Vijay-1 commented:
@mitchell852 
Ok. Have to get the required GoLang code inside leaving the API (GoLang) part for now. Will try to get this done. 

@rawlinp 
I will address this and let you know.

Vijay-1 commented on traffic_ops/app/db/migrations/20180521000000_cache_group_fallback.sql:
```diff
@@ -0,0 +1,38 @@
+
+-- cachegroup_fallbacks
+CREATE TABLE cachegroup_fallbacks (
+    primary_cg bigint,
```
There wont be any Database OP with these values as Null. Do we really need to do this?

Vijay-1 commented on traffic_ops/app/db/migrations/20180521000000_cache_group_fallback.sql:
```diff
@@ -0,0 +1,38 @@
+    UNIQUE (primary_cg, set_order)
+); 
+
+ALTER TABLE cachegroup ADD COLUMN fallback_to_closest BOOLEAN DEFAULT TRUE;
```
NOT NULL may fail any DB Ops if the User doesnt supply fallback_to_closest. Also, this is not a mandatory field by design and hence i think we dont have to change this.

Vijay-1 commented on traffic_ops/app/lib/API/CachegroupFallback.pm:
```diff
@@ -0,0 +1,278 @@
+		$cache_id = $param_array[0]{cacheGroupId};
+	}
+
+	if ( !defined($params) ) {
```
Yes. Will address.

Vijay-1 commented on traffic_ops/app/lib/API/Cachegroup.pm:
```diff
@@ -57,6 +57,7 @@ sub index {
 				"lastUpdated"                   => $row->last_updated,
 				"parentCachegroupId"            => $row->parent_cachegroup_id,
 				"parentCachegroupName"          => ( defined $row->parent_cachegroup_id ) ? $idnames{ $row->parent_cachegroup_id } : undef,
+				"fallbackToClosest"           => \$row->fallback_to_closest,
```
Addressed.

rawlinp commented on traffic_ops/traffic_ops_golang/crconfig/edgelocations.go:
```diff
@@ -57,6 +57,36 @@ and (st.name = 'REPORTED' or st.name = 'ONLINE' or st.name = 'ADMIN_DOWN')
 		if ttype == RouterTypeName {
 			routerLocs[cachegroup] = latlon
 		} else {
+			primaryCacheId := ""
```
rather than re-querying the cachegroup's ID here, you should just update the query on L35 to also select `cg.id` and scan that into a variable.

In fact, you might even be able to eliminate the query on L65 altogether by modifying the query on L35 to include a subquery. Something like:
```
select cg.name, ..., ARRAY(select c.name from cachegroup_fallbacks cgf join cachegroup c on c.id = cgf.backup_cg where cgf.primary_cg = cg.id) ... from ...
```
I haven't tried that query out, but I think it's worth a shot rather than running an extra query per cachegroup. If it's possible, you should be able to scan that result into a slice of strings and just iterate over it.

rawlinp commented on traffic_ops/traffic_ops_golang/crconfig/edgelocations.go:
```diff
@@ -57,6 +57,36 @@ and (st.name = 'REPORTED' or st.name = 'ONLINE' or st.name = 'ADMIN_DOWN')
+				return nil, nil, errors.New("Failed while retrieving from cachegroup: " + err.Error())
+			}
+
+			dbRows, err := db.Query(`select backup_cg from cachegroup_fallbacks where primary_cg = $1 order by set_order`, primaryCacheId)
```
you can use a `join` here and remove the extra query on L79:
```
select cgf.backup_cg, c.name from cachegroup_fallbacks cgf join cachegroup c on c.id = cgf.backup_cg
```

rawlinp commented on traffic_ops/app/db/migrations/20180521000000_cache_group_fallback.sql:
```diff
@@ -0,0 +1,38 @@
+/*
```
this file will need renamed again because another migration was recently merged

rawlinp commented on traffic_ops/traffic_ops_golang/crconfig/edgelocations.go:
```diff
@@ -57,6 +57,36 @@ and (st.name = 'REPORTED' or st.name = 'ONLINE' or st.name = 'ADMIN_DOWN')
+
+			index := 0
+			for dbRows.Next() {
+				backup_id := ""
```
in golang the typical naming style is to use camelcase, i.e. `backupID`

Vijay-1 commented on lib/go-tc/cachegroupfallback.go:
```diff
@@ -0,0 +1,55 @@
+// A List of CACHEGROUPFALLBACKs Response
+// swagger:response CACHEGROUPFALLBACKsResponse
+// in: body
+type CACHEGROUPFALLBACKsResponse struct {
```
Ok. Will change this.

Vijay-1 commented on traffic_ops/app/lib/API/CachegroupFallback.pm:
```diff
@@ -0,0 +1,287 @@
+	if ( ($rs_backups->count > 0) ) {
+		my $del_records = $rs_backups->delete();
+		if ($del_records) {
+			&log( $self, "Backup configuration DELETED", "APICHANGE");
```
Right now, we can delete fallbacks using primary ID, backupID and a combination of both and that is the reason behind leaving the IDs. Is it ok , i code an if-else block whose purpose is to assist only in logging?

rawlinp replied:
For 1, yeah I think adding logic so that we get an audit record of what was actually deleted is necessary. Otherwise this log provides no value.

rawlinp replied:
For 2, I'm ok with leaving as-is in the Perl, but in the Golang API it will be wrapped in a transaction so that all or none of the requested backups are created/updated. People should not have to double-check our API when getting a 200 success code.

For 3 let's just keep it nullable for now.

Vijay-1 commented on traffic_ops/app/lib/API/CachegroupFallback.pm:
```diff
@@ -0,0 +1,287 @@
+	foreach my $config (@{ $params }) {
+		my $rs_backup = $self->db->resultset('Cachegroup')->search( { id => $config->{fallbackId} } )->single();
+		if ( !defined($rs_backup) ) {
+			&log( $self, "ERROR Backup config: No such Cache Group $config->{fallbackId}");
```
Ok. I have a doubt on this comment. Say i get a list of 10 fallbacks, where the 10th one is invalid. when i reach the 10th, i have already inserted the other 9. That is the reason i coded it this way so that only the valid elements will be inserted. Also , on a successful POST / PUT, the user can always do a GET to verify.

rawlinp replied:
If the client receives a 200, they should expect that the entire request was successful and shouldn't have to verify that all backups were actually created. Thinking about how we'd do this in the golang API, the entire thing would likely be wrapped in a transaction so that it creates either none or all of the requested backups. I'm ok with leaving this in the Perl as-is but definitely think it should be entirely wrapped in a transaction in the golang API.

rawlinp commented on traffic_ops/traffic_ops_golang/crconfig/edgelocations.go:
```diff
@@ -49,20 +49,20 @@ and (st.name = 'REPORTED' or st.name = 'ONLINE' or st.name = 'ADMIN_DOWN')
 
 	for rows.Next() {
 		cachegroup := ""
+		primaryCacheID := ""
```
kind of a nitpick, but I think this should be `primaryCacheID := 0` to match the int DB type

Vijay-1 replied:
Addressed.

Vijay-1 commented on lib/go-tc/cachegroupfallback.go:
```diff
@@ -20,33 +20,33 @@ package tc
 
-// A List of CACHEGROUPFALLBACKs Response
-// swagger:response CACHEGROUPFALLBACKsResponse
+// A List of cachegroupFallbacks Response
```
Addressed.

Vijay-1 commented:
@rawlinp  rebased

rawlinp commented on traffic_ops/traffic_ops_golang/crconfig/edgelocations.go:
```diff
@@ -32,7 +32,7 @@ func makeLocations(cdn string, db *sql.DB) (map[string]tc.CRConfigLatitudeLongit
 	// TODO test whether it's faster to do a single query, joining lat/lon into servers
 	q := `
-select cg.name, t.name as type, cg.latitude, cg.longitude from cachegroup as cg
+select cg.name, cg.id, t.name as type, cg.latitude, cg.longitude from cachegroup as cg
```
You need to select cg.fallback_to_closest here and scan the value into a `var fallbackToClosest *bool`. If it's null, set `BackupLocations.FallbackToClosest` below to false. Otherwise, set it to the dereferenced value.

Also, I noticed the CRConfig generated sets the value of `fallbackToClosest` to a string rather than a boolean. Since booleans are an actual type in Go, I'm thinking we should set the values to booleans rather than strings and make sure Traffic Router handles that properly.

Vijay-1 replied:
I took a look and found a bunch of bools which gets written as string in this particular part. For example=> sslEnabled.

rawlinp replied:
I think the string boolean thing might have stemmed from Perl not having real booleans. Now that it's been migrated to Go, we _could_ use real booleans in the JSON. But since there are a bunch of string bools in the CRConfig already, I guess it's better to be consistent and keep it as a string bool.

rawlinp commented on traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java:
```diff
@@ -716,17 +722,44 @@ public CacheLocation getCoverageZoneCacheLocation(final String ip, final String
+		if (closestCacheLocation != null) {
+			LOGGER.debug("Got closest CZ cache group " + closestCacheLocation.getId() + " for " + ip + ", ds " + deliveryServiceId);
+			if (track != null) {
+				track.setFromBackupCzGroup(true);
```
Why do we set this to true here? If we reach this point, there were no available backup cachegroups (or none configured), so we shouldn't track it as from a backup cachegroup, right?

Vijay-1 replied:
The cache which gets selected in this cache is not a primary one. This is again a backup cache configured via Coordinates and hence it gets tracked as backup.

rawlinp replied:
Ok, I can agree with that. It would be useful to track those occurrences as "backup hits".

rawlinp commented:
Finished testing the TR changes; they still look good except for my question about the tracking.

rawlinp commented on traffic_ops/traffic_ops_golang/crconfig/edgelocations.go:
```diff
@@ -49,14 +49,49 @@ and (st.name = 'REPORTED' or st.name = 'ONLINE' or st.name = 'ADMIN_DOWN')
 		} else {
+			q := `select cachegroup.name from cachegroup_fallbacks
+join cachegroup on cachegroup_fallbacks.backup_cg = cachegroup.id
+and cachegroup_fallbacks.primary_cg = $1
```
I missed this in my previous review, but this query is missing `order by cachegroup_fallbacks.set_order`. Without this the `set_order` ordering is not enforced.

Vijay-1 replied:
Right. I missed that. Addressed.

Vijay-1 commented:
Yes. The changes requested for edgeLocations.go for CRConfig were addressed.
